### PR TITLE
[Malleability Immutable] Removed non-constructor mutations for TimeoutObject and TimeoutCertificate

### PR DESCRIPTION
**Pull request by UlyanaAndrukhiv**
Closes: #7272, #7302

## Context

In this PR :
- added constructors for `TimeoutObject` and `TimeoutCertificate` structs,
- removed non-constructor mutations for those structs,
- added `UntrustedTimeoutObject` and `UntrustedTimeoutCertificate` models,
- added unit tests for new constructors.
- fixed flaky `TestNewEpochStateContainer` test (part with unsorted identities)

### CODE REVIEW

**Comment by UlyanaAndrukhiv** [comment]
@jordanschalm , @tim-barry, this PR is updated and ready for review.

**Comment by UlyanaAndrukhiv** [comment]
@jordanschalm, I want to ask about scenario which will mutate a struct `TimeoutCertificate`by using `atomic.UnsafePointer`. Should we also use constructor in such case?

https://github.com/onflow/flow-go/blob/8c30c496cf7b6e3fadad97328697a42f633a4853/consensus/hotstuff/tracker/tracker.go#L55 
https://github.com/onflow/flow-go/blob/8c30c496cf7b6e3fadad97328697a42f633a4853/consensus/hotstuff/tracker/tracker.go#L79

**Comment by jordanschalm** [comment]
> @jordanschalm, I want to ask about scenario which will mutate a struct TimeoutCertificateby using atomic.UnsafePointer. Should we also use constructor in such case?

Good question. I don't think we need to use a constructor here to maintain the mutation-free properties we want. To me this is conceptually closer to re-assignment of a pointer variable than mutation of a TC. (`NewestTCTracker` mutates the pointer variable `newestTC` so it points to different data, but doesn't mutate the data it points to.)

File: engine/execution/messages.go [hunk 1]
```diff
@@ -1,17 +1,54 @@
...
+	"fmt"
+
 	"github.com/onflow/flow-go/model/flow"
 	"github.com/onflow/flow-go/module/mempool/entity"
 )
 
+// ComputationResult captures artifacts of execution of block collections, collection attestation results and
+// the full execution receipt, as sent by the Execution Node.
+//
+//structwrite:immutable - mutations allowed only within the constructor
```
Comment (jordanschalm) [inline engine/execution/messages.go]:
```suggestion
// CAUTION: This type is used to represent both a complete ComputationResult and a partially constructed ComputationResult.
// TODO: Consider using a Builder type to represent the partially constructed model.
```

File: engine/execution/messages.go [hunk 2]
```diff
@@ -1,17 +1,54 @@
...
+//
+// An instance of UntrustedComputationResult should be validated and converted into
+// a trusted ComputationResult using NewComputationResult constructor.
+type UntrustedComputationResult ComputationResult
+
+// NewComputationResult creates a new instance of ComputationResult.
+// Construction ComputationResult allowed only within the constructor.
+//
+// All errors indicate a valid ComputationResult cannot be constructed from the input.
+func NewComputationResult(untrusted UntrustedComputationResult) (*ComputationResult, error) {
```
Comment (jordanschalm) [inline engine/execution/messages.go]: As discussed at standup today, there are a few issues with `ComputationResult` malleability and immutability.

First, it has an ID method promoted through embedding `ExecutionReceipt`. However that ID method is trivially not collision resistant (any mutation to `BlockExecutionResult` and `BlockAttestationResult` results in the same ID.
- That means `ComputationResult` is currently malleable.
- This is also why the issue for making `ComputationResult` immutable was created (because it implements `flow.Entity`, which is how the malleability issue set was created).

Second, it remains mutable with this PR, because the embedded types are not also marked as immutable with `structwrite`. For example, [AppendCollectionAttestationResult](https://github.com/onflow/flow-go/blob/c12d83f06a63f8cc552764a89bb4269eb7da2573/engine/execution/block_result.go#L176) will mutate `ComputationResult`. (In general, if we have a deeply nested struct and we want the whole thing to be immutable, we need to opt in all the constituent struct types to the `structwrite` linter -- in this case `BlockExecutionResult` and `BlockAttestationResult`.)

Anyway, I should not have made an Immutability issue for this type (sorry about that). It snuck in as an edge case in the way the list of types was constructed. Really what we want is to remove its implementation of `flow.Entity`; it does not have a canonical ID and it is not sent over the network.

### Suggested Changes
- Remove the constructor and tests for `ComputationResult`
- Ensure `ComputationResult` does not accidentally inherit a non-collision-resistant ID method. I took a look and it seems like the following two steps will accomplish this with minimal additional changes:
  - Change `ExecutableBlock.ID` to `ExecutableBlock.BlockID` (already done on `feature/malleability`)
  - Make `ComputationResult.ExecutionReceipt` non-embedded: 
```
type ComputationResult struct {
	*BlockExecutionResult
	*BlockAttestationResult

	ExecutionReceipt *flow.ExecutionReceipt
}
```

Comment (UlyanaAndrukhiv) [reply engine/execution/messages.go]: Thanks for the detailed explanation! I'll remove all changes related to `ComputationResult` from this PR and create a separate pull request to `feature/malleability` to make `ComputationResult.ExecutionReceipt` non-embedded, since `ExecutableBlock.ID` has already been changed to `ExecutableBlock.BlockID` there.

Comment (UlyanaAndrukhiv) [reply engine/execution/messages.go]: Suggested changes implemented in PR #7511

**Review by jordanschalm — APPROVED** [review]

File: model/flow/timeout_certificate.go [hunk 3]
```diff
@@ -21,6 +27,63 @@ type TimeoutCertificate struct {
...
+//
+// All errors indicate a valid TimeoutCertificate cannot be constructed from the input.
+func NewTimeoutCertificate(untrusted UntrustedTimeoutCertificate) (*TimeoutCertificate, error) {
+	if untrusted.NewestQC == nil {
+		return nil, fmt.Errorf("newest QC must not be nil")
+	}
+	if untrusted.SignerIndices == nil {
+		return nil, fmt.Errorf("signer indices must not be nil")
+	}
+	if untrusted.SigData == nil {
```
Comment (durkmurder) [inline model/flow/timeout_certificate.go]:
```suggestion
	if len(untrusted.SigData) == 0 {
```

File: model/flow/timeout_certificate.go [hunk 4]
```diff
@@ -21,6 +27,63 @@ type TimeoutCertificate struct {
...
+
+// NewTimeoutCertificate creates a new instance of TimeoutCertificate.
+// Construction TimeoutCertificate allowed only within the constructor.
+//
+// All errors indicate a valid TimeoutCertificate cannot be constructed from the input.
+func NewTimeoutCertificate(untrusted UntrustedTimeoutCertificate) (*TimeoutCertificate, error) {
+	if untrusted.NewestQC == nil {
+		return nil, fmt.Errorf("newest QC must not be nil")
+	}
+	if untrusted.SignerIndices == nil {
```
Comment (durkmurder) [inline model/flow/timeout_certificate.go]:
```suggestion
	if len(untrusted.SignerIndices) == 0 {
```

File: model/flow/timeout_certificate.go [hunk 5]
```diff
@@ -21,6 +27,63 @@ type TimeoutCertificate struct {
...
+	// Note: we specifically allow for the TC to have the same view as the highest QC.
+	// This is useful as a fallback, because it allows replicas other than the designated
+	// leader to also collect votes and generate a QC.
+	if untrusted.View < untrusted.NewestQC.View {
+		return nil, fmt.Errorf("TC's QC view (%d) cannot be newer than the TC's view (%d)", untrusted.NewestQC.View, untrusted.View)
+	}
+
+	// verifying that tc.NewestQC is the QC with the highest view.
+	// Note: A byzantine TC could include `nil` for tc.NewestQCViews
+	if untrusted.NewestQCViews != nil {
```
Comment (durkmurder) [inline model/flow/timeout_certificate.go]:
```suggestion
	if len(untrusted.NewestQCViews) > 0 {
```

Additionally, a TC with empty `NewestQCViews` is considered invalid. Moreover `len(NewestQCViews)` == `len(signers)` but we have no easy way to extract singers without parsing singer indices.

File: model/flow/timeout_certificate_test.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+					helper.WithTCView(99),
+					helper.WithTCNewestQC(qc),
+				)),
+		)
+		require.Error(t, err)
+		require.Nil(t, tc)
+		assert.Contains(t, err.Error(), "TC's QC view (100) cannot be newer than the TC's view (99)")
+	})
+
+	t.Run("invalid input when NewestQCViews has view higher than NewestQC.View", func(t *testing.T) {
```
Comment (durkmurder) [inline model/flow/timeout_certificate_test.go:141]: can you add a test case when `NewestQCViews` is empty(both nil and empty)?

File: model/flow/timeout_certificate_test.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+		tc := helper.MakeTC()
+		tc.SignerIndices = nil
+
+		res, err := flow.NewTimeoutCertificate(flow.UntrustedTimeoutCertificate(*tc))
+		require.Error(t, err)
+		require.Nil(t, res)
+		assert.Contains(t, err.Error(), "signer indices must not be nil")
+	})
+
+	t.Run("invalid input with nil SigData", func(t *testing.T) {
```
Comment (durkmurder) [inline model/flow/timeout_certificate_test.go:87]: I would suggest to expand it and check both for nil and empty

File: model/flow/timeout_certificate_test.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+		tc := helper.MakeTC()
+		tc.NewestQC = nil
+
+		res, err := flow.NewTimeoutCertificate(flow.UntrustedTimeoutCertificate(*tc))
+		require.Error(t, err)
+		require.Nil(t, res)
+		assert.Contains(t, err.Error(), "newest QC must not be nil")
+	})
+
+	t.Run("invalid input with nil SignerIndices", func(t *testing.T) {
```
Comment (durkmurder) [inline model/flow/timeout_certificate_test.go:67]: I would suggest to expand it and check both for nil and empty

File: consensus/hotstuff/model/timeout.go [hunk 9]
```diff
@@ -61,6 +63,63 @@ type TimeoutObject struct {
...
+	// If no QC is included, we require a TC to be present, which by check must be for
+	// the previous round.
+	lastViewSuccessful := untrusted.View == untrusted.NewestQC.View+1
+	if !lastViewSuccessful {
+		// The TO's sender did _not_ observe a QC for round timeout.View-1. Hence, it should
+		// include a TC for the previous round. Otherwise, the TO is invalid.
+		if untrusted.LastViewTC == nil {
+			return nil, fmt.Errorf("must include TC")
+		}
+	}
```
Comment (durkmurder) [inline consensus/hotstuff/model/timeout.go:117]: We can also check the sigdata and fail early if it's empty

Comment (durkmurder) [reply consensus/hotstuff/model/timeout.go:117]: same goes for `SignerID`

File: model/flow/timeout_certificate_test.go (hunk 8 above)
Comment (UlyanaAndrukhiv) [reply model/flow/timeout_certificate_test.go:67]: Added in [8dae2ff](https://github.com/onflow/flow-go/pull/7433/commits/8dae2ff3d8893a7ac916d7349b8a96e2b0f75af2)

File: model/flow/timeout_certificate_test.go (hunk 7 above)
Comment (UlyanaAndrukhiv) [reply model/flow/timeout_certificate_test.go:87]: Added in [8dae2ff](https://github.com/onflow/flow-go/pull/7433/commits/8dae2ff3d8893a7ac916d7349b8a96e2b0f75af2)

File: model/flow/timeout_certificate_test.go (hunk 6 above)
Comment (UlyanaAndrukhiv) [reply model/flow/timeout_certificate_test.go:141]: Added in [fb91871](https://github.com/onflow/flow-go/pull/7433/commits/fb918714bc0ea4e430ec984675d462b96020a396)

File: consensus/hotstuff/model/timeout.go (hunk 9 above)
Comment (UlyanaAndrukhiv) [reply consensus/hotstuff/model/timeout.go:117]: Added in [77d3aff](https://github.com/onflow/flow-go/pull/7433/commits/77d3affe51f397a03fe95fba7bec8d0af6e20df1)

**Comment by UlyanaAndrukhiv** [comment]
@durkmurder, thank you for all the suggestions! The PR has been updated and is ready for the second round of review.

File: consensus/hotstuff/model/timeout.go [hunk 10]
```diff
@@ -61,6 +63,70 @@ type TimeoutObject struct {
...
+	if untrusted.NewestQC == nil {
+		return nil, fmt.Errorf("newest QC must not be nil")
+	}
+	if untrusted.SignerID == flow.ZeroID {
+		return nil, fmt.Errorf("signer ID must not be zero")
+	}
+	if len(untrusted.SigData) == 0 {
+		return nil, fmt.Errorf("signature must not be empty")
+	}
+
```
Comment (durkmurder) [inline consensus/hotstuff/model/timeout.go]:
```suggestion
```

File: model/flow/timeout_certificate.go [hunk 11]
```diff
@@ -21,6 +27,65 @@ type TimeoutCertificate struct {
...
+			if newestQCView < view {
+				newestQCView = view
+			}
+		}
+		if newestQCView > untrusted.NewestQC.View {
+			return nil, fmt.Errorf("included QC (view=%d) should be equal or higher to highest contributed view: %d", untrusted.NewestQC.View, newestQCView)
+		}
+	} else {
+		return nil, fmt.Errorf("newest QC views must not be empty")
+	}
```
Comment (durkmurder) [inline model/flow/timeout_certificate.go]:
```suggestion
	if len(untrusted.NewestQCViews) == 0 {
		return nil, fmt.Errorf("newest QC views must not be empty")
	}
	
	newestQCView := untrusted.NewestQCViews[0]
	for _, view := range untrusted.NewestQCViews {
		if newestQCView < view {
			newestQCView = view
		}
	}
	if newestQCView > untrusted.NewestQC.View {
		return nil, fmt.Errorf("included QC (view=%d) should be equal or higher to highest contributed view: %d", untrusted.NewestQC.View, newestQCView)
	}
```
I would just inverse it and save a nesting.
